File: 2022/go/day8/day8.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	input, _ := os.ReadFile("../../input/day8.txt")
	heights := parseMap(string(input))

	fmt.Printf("part1 : %d\n", countVisible(heights))
	fmt.Printf("part2 : %d\n", highestScenicScore(heights))
}

func highestScenicScore(heights [][]byte) (max int) {
	for y, row := range heights {
		for x := range row {
			score := scenicScore(heights, y, x)
			if score > max {
				max = score
			}
		}
	}
	return max
}

func scenicScore(heights [][]byte, y, x int) (score int) {
	up := visibleTrees(heights, y, x, func(x, y int) (int, int) { return x, y - 1 })
	down := visibleTrees(heights, y, x, func(x, y int) (int, int) { return x, y + 1 })
	left := visibleTrees(heights, y, x, func(x, y int) (int, int) { return x - 1, y })
	right := visibleTrees(heights, y, x, func(x, y int) (int, int) { return x + 1, y })
	return up * down * left * right
}

func visibleTrees(heights [][]byte, oy, ox int, move func(x, y int) (int, int)) (count int) {
	for x, y := move(ox, oy); inside(heights, y, x); x, y = move(x, y) {
		count++
		if heights[y][x] >= heights[oy][ox] {
			return count
		}
	}
	return count
}

func inside(heights [][]byte, y, x int) bool {
	return y >= 0 && y < len(heights) && x >= 0 && x < len(heights[y])
}

func countVisible(heights [][]byte) (sum int) {
	for y, row := range heights {
		for x := range row {
			if visible(heights, y, x) {
				sum++
			}
		}
	}
	return sum
}

func visible(heights [][]byte, y, x int) bool {
	if y == 0 || y == len(heights)-1 || x == 0 || x == len(heights[y])-1 {
		return true // edge cases
	}
	height := heights[y][x]
	row := heights[y]

	if max(row[:x]) < height || max(row[x+1:]) < height {
		return true
	}
	col := column(heights, x)
	if max(col[:y]) < height || max(col[y+1:]) < height {
		return true
	}
	return false
}

func column(heights [][]byte, x int) (col []byte) {
	for y := 0; y < len(heights); y++ {
		col = append(col, heights[y][x])
	}
	return col
}

func max(data []byte) (max byte) {
	for _, b := range data {
		if b > max {
			max = b
		}
	}
	return max
}

func parseMap(input string) (heights [][]byte) {
	lines := strings.Split(input, "\n")
	for _, line := range lines {
		var row []byte
		for _, c := range line {
			row = append(row, byte(c-'0'))
		}
		heights = append(heights, row)
	}
	return heights
}

File: 2022/java/Day8.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.function.Function;
import java.util.stream.IntStream;
import java.util.stream.Stream;

class Day8 {

    public static void main(String ... args) throws Exception {
        var map = parseMap(Files.readString(Path.of("../input/day8.txt")));

        System.out.printf("part 1: %s\n", countVisible(map));
        System.out.printf("part 2: %s\n", highestScenicScore(map));
    }

    static int highestScenicScore(int[][] map) {
        return IntStream.range(0, map.length)
            .flatMap(y -> IntStream.range(0, map[y].length)
                .map(x -> scenicScore(map, new Point(x, y)))).max().getAsInt();
    }

    static int scenicScore(int[][] map, Point op) {
        var up = visibleTrees(map, op, p -> p.move(0, -1));
        var down = visibleTrees(map, op, p -> p.move(0, 1));
        var left = visibleTrees(map, op, p -> p.move(-1, 0));
        var right = visibleTrees(map, op, p -> p.move(1, 0));

        return up * down * left * right;
    }

    static int visibleTrees(int[][] map, Point op, Function<Point,Point> move) {
        int count = 0;
        for (var p = move.apply(op); p.inside(map); p = move.apply(p)) {
            count++;
            if (map[p.y()][p.x()] >= map[op.y()][op.x()]) {
                return count;
            }
        }
        return count;
    }

    static long countVisible(int[][] map) {
        return IntStream.range(0, map.length)
            .flatMap(y -> IntStream.range(0, map[y].length).filter(x -> visible(map, y, x))).count();
    }

    static boolean visible(int[][] map, int y, int x) {
        if (y == 0 || y == map.length-1 || x == 0 || x == map[y].length-1) {
            return true; // edge cases
        }
        var height = map[y][x];
        return Stream.of(
            IntStream.of(map[y]).limit(x).max(), // left max height
            IntStream.of(map[y]).skip(x+1).max(), // right max
            IntStream.of(col(map,x)).limit(y).max(), // up
            IntStream.of(col(map,x)).skip(y+1).max()) // down
            .mapToInt(m -> m.getAsInt()).anyMatch(h -> h < height); // any shorter than given tree
    }

    static int[] col(int[][] map, int x) {
        return IntStream.range(0, map.length).map(y -> map[y][x]).toArray();
    }

    static int[][] parseMap(String input) {
        return input.lines().map(l -> l.chars().map(c -> c - '0').toArray()).toArray(int[][]::new);
    }
}

record Point(int x, int y) {
    Point move(int mx, int my) { return new Point(x + mx, y + my);}
    boolean inside(int[][] map) { return y >= 0 && y < map.length && x >= 0 && x < map[y].length; }
}